system/lib/libc/dynlink.c

```diff
@@ -105,7 +108,241 @@ static struct dso* load_library_start(const char* name, int flags) {
+  // with sync_one_onerror.
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  sync_next(thread_local_tail->next, promise_id);
+}
+
+// Called on background threads to synchronously "catch up" with all the DSOs
+// that are currently loaded.
+bool _emscripten_thread_sync_code() {
+  // Should only ever be called from a background thread.
+  assert(!emscripten_is_main_runtime_thread());
+  ensure_init();
+  if (thread_local_tail == tail) {
+#ifdef DYLINK_DEBUG
+    _emscripten_errf("%p: emscripten_thread_sync_code: already in sync", pthread_self());
+#endif
+    return true;
+  }
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  while (thread_local_tail->next) {
+    struct dso* p = thread_local_tail->next;
+#ifdef DYLINK_DEBUG
+    _emscripten_errf(
+      "%p: emscripten_thread_sync_code: %s mem_addr=%p mem_size=%zu "
+      "table_addr=%p table_size=%zu",
+      pthread_self(),
+      p->name,
+      p->mem_addr,
+      p->mem_size,
+      p->table_addr,
+      p->table_size);
+#endif
+    void* success = _dlopen_js(p);
+    if (!success) {
+      pthread_rwlock_unlock(&lock);
+      _emscripten_errf("%p: emscripten_thread_sync_code failed: %s", pthread_self(), dlerror());
+      return false;
+    }
+    thread_local_tail = p;
+  }
+  pthread_rwlock_unlock(&lock);
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: emscripten_thread_sync_code done", pthread_self());
+#endif
+  return true;
+}
+
+// This structure exists so that we can stash the status of work we do
+// during emscripten_proxy_async_with_callback.
+// If the function that emscripten_proxy_async_with_callback ran included
+// a return value we could avoid this struct and the malloc/free of it.
+// See https://github.com/emscripten-core/emscripten/issues/18378
+struct thread_sync_data {
+  int promise_id;
+  bool result;
+};
+
+static void do_thread_sync(void* arg) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: do_thread_sync: %d", pthread_self(), data->promise_id);
+#endif
+  data->result = _emscripten_thread_sync_code();
+}
+
+// Called once _emscripten_proxy_sync_code completes
+static void done_thread_sync(void* arg) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: done_thread_sync: promise_id=%d success=%d",
+                   pthread_self(),
+                   data->promise_id,
+                   data->result);
+#endif
+  if (data->result) {
+    _emscripten_promise_resolve(data->promise_id, NULL);
+  } else {
+#if ABORT_ON_SYNC_FAILURE
+    abort();
+#else
+    _emscripten_promise_reject(data->promise_id);
+#endif
+  }
+  free(data);
+}
+
+// Proxying queue specically for handling code loading (dlopen) events.
+// Initialized on first call to `_emscripten_proxy_sync_code below.
+static em_proxying_queue * _Atomic dlopen_proxying_queue = NULL;
```

Comment:
Why does this have to be atomic? Aren't all accesses to it protected by locks?

Reply:
On the send end yes.. but all threads constantly check the queue in `_emscripten_proxy_sync_code`.  This is called by `emscripten_yield` which itself is called by `futex_wait`.

I chose to only create the queue when a item is first placed into it.. this means that `_emscripten_proxy_sync_code` is basically doing nothing until the first call to `dlopen`.

If I was to create the queue in `_emscripten_proxy_sync_code` it would need some kind of lock around it.. and it would also never get delayed so it might as well be a static ctor in that case.

Reply:
Makes sense, thanks! It would be good to add a code comment about the locking strategy.